File: 20160113-StringCalculator-Java8/src/main/java/kata/SC.java
package kata;

import java.util.regex.Pattern;
import java.util.stream.Stream;

import static java.util.regex.Pattern.quote;
import static java.util.stream.Collectors.joining;

public class SC {
    public static int add(String input) {
//        System.out.println("**[" + input + "]**");

        if (parse(input).anyMatch(n -> n < 0)) {
            throw new IllegalArgumentException(parse(input).filter(n -> n < 0).map(Object::toString).collect(joining(",")));
        } else {
            return sum(parse(input).filter(n -> n < 1001));
        }
    }

    public static Stream<Integer> parse(String input) {
        if (input.isEmpty()) {
            return tokenize("0", ",");
        } else if (input.startsWith("//[")) {
            final int indexOfNewline = input.indexOf('\n');
            final String numbers = input.substring(indexOfNewline + 1);
            final String separators = input.substring(3, indexOfNewline - 1);

            return tokenize(numbers, Stream.of(separators.split("\\]\\[")).sorted((x, y) -> y.length() - x.length()).map(Pattern::quote).collect(joining("|")));
        } else if (input.startsWith("//")) {
            final String numbers = input.substring(4);
            final String separator = input.substring(2, 3);

            return tokenize(numbers, quote(separator));
        } else {
            return tokenize(input, ",|\n");
        }
    }

    private static Stream<Integer> tokenize(String numbers, String regexSeparator) {
        return Stream.of(numbers.split(regexSeparator)).map(Integer::parseInt);
    }

    private static int sum(Stream<Integer> numbers) {
        return numbers.reduce(0, (accumulator, n) -> accumulator + n);
    }
}

File: 20160425-StringCalculator-Kotlin/src/main/kotlin/kata/SC.kt
package kata

import java.util.regex.Pattern.quote as quoteRegex

fun add(input: String): Int {
    val numbers = parse(input)

    if (numbers.any { it < 0 })
        throw IllegalArgumentException(numbers.filter { it < 0 }.joinToString(","))
    else
        return numbers.filter { it < 1001 }.fold(0) { sum, item -> sum + item }
}

private fun parse(input: String): List<Int> = tokenize(input).map { Integer.parseInt(it) }

private fun tokenize(input: String): List<String> =
        when {
            input.isEmpty() -> listOf()
            input.startsWith("//[") -> {
                val indexOfNewline = input.indexOf("\n")
                val numbers = input.substring(indexOfNewline + 1)
                val separator = input.substring(3, indexOfNewline - 1).split("][").map { quoteRegex(it) }.joinToString("|").toRegex()

                numbers.split(separator)
            }
            input.startsWith("//") -> input.substring(4).split(input.substring(2, 3))
            else -> input.split(",|\n".toRegex())
        }
